batch/batch.go

```diff
@@ -72,12 +67,23 @@ func (m *batchManager) loadExistingBatches() error {
 		return fmt.Errorf("loadExistingBatches: retrieve batches: %v", err)
 	}
 	for idx := range vals {
-		batch, err := m.newBatch(vals[idx], &batchMeta{})
-		if err != nil {
+		m.mu.Lock()
+		batch := &batch{
+			Id:       vals[idx],
+			Parents:  make([]*batch, 0),
+			Children: make([]*batch, 0),
+			Meta:     &batchMeta{},
+		}
+
+		if err := m.loadMetadata(batch); err != nil {
 			util.Warnf("loadExistingBatches: error load batch (%s) %v", vals[idx], err)
+			m.remove(batch)
+			m.mu.Unlock()
 			continue
 		}
 		m.Batches[vals[idx]] = batch
+		m.batchMu[vals[idx]] = &sync.Mutex{}
+		m.mu.Unlock()
```

Comment:
I think if there's a panic here (for some reason) a lock could never be released. What about extracting this into a separate function so you can do this:
```
func loadBatch(...) {
m.mu.lock()
defer m.mu.Unlock()
...
}
```

Reply:
eventually in the chain we call `m.getBatch` which also calls `m.mu.lock()` which will create a dead lock. We only use the lock when accessing `m.Batches`